sound/soc/sof/sof-client-ipc-kernel-injector.c

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2023 Google Inc. All rights reserved.
+//
+// Author: Curtis Malainey <cujomalainey@chromium.org>
+//
+
+#include <linux/auxiliary_bus.h>
+#include <linux/debugfs.h>
+#include <linux/mod_devicetable.h>
+#include <linux/module.h>
+#include <linux/pm_runtime.h>
+#include <sound/sof/header.h>
+
+#include "sof-client.h"
+
+#define SOF_IPC_CLIENT_SUSPEND_DELAY_MS	3000
+
+struct sof_msg_inject_priv {
+	struct dentry *kernel_dfs_file;
+	size_t max_msg_size;
+
+	void *kernel_buffer;
+};
+
+static int sof_msg_inject_dfs_open(struct inode *inode, struct file *file)
+{
+	int ret = debugfs_file_get(file->f_path.dentry);
+
+	if (unlikely(ret))
+		return ret;
+
+	ret = simple_open(inode, file);
+	if (ret)
+		debugfs_file_put(file->f_path.dentry);
+
+	return ret;
+}
+
+static ssize_t sof_kernel_msg_inject_dfs_write(struct file *file, const char __user *buffer,
+					       size_t count, loff_t *ppos)
+{
+	struct sof_client_dev *cdev = file->private_data;
+	struct sof_msg_inject_priv *priv = cdev->data;
+	struct sof_ipc_cmd_hdr *hdr = priv->kernel_buffer;
+	ssize_t size;
+
+	if (*ppos)
+		return 0;
+
+	size = simple_write_to_buffer(priv->kernel_buffer, priv->max_msg_size,
+				      ppos, buffer, count);
+	if (size < 0)
+		return size;
+	if (size != count)
+		return -EFAULT;
+
+	size = sof_client_ipc_rx_message(cdev, hdr, priv->kernel_buffer);
+	if (size)
+		return size;
```

Comment:
you also want a pm_runtime_resume_and_get() before sending an IPC and a pm_runtime_mark_last_busy() + pm_runtime_put_autosuspend() after the IPC?

See e.g. the SoundWire example here: https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/debugfs.c#L59 

https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/debugfs.c#L123

Reply:
actually do we need PM? This fs is not interacting with the HW in anything way.

Reply:
if you application that writes into debugfs lets the DSP suspend, you'll never get any IPC sent. Somehow you need to resume the DSP then send the IPC.

The client to inject the message is a child of the PCI device, so if you pm_runtime_resume it the PCI is also resumed. Conversely the PCI device will only suspend if you application stops sending messages for more than 3s.

Reply:
The IPC isn't going into the DSP though. Its going back into the kernel. The IPC is being sent falsely from the DSP to provide an interface to fuzz the kernel

Reply:
what happens if the driver wants to dump a state or register which isn't powered?

Notifications from the DSP can only come when the DSP is powered, I would err on the side of avoiding to add more variables...

Reply:
Good point, we should force the DSP awake if we are going to pretend to be it as we won't know what we will trigger